aws-opentelemetry-distro/src/amazon/opentelemetry/distro/exporter/aws/metrics/aws_cloudwatch_emf_exporter.py

```diff
@@ -0,0 +1,523 @@
+                if data_point.time_unix_nano is not None
+                else int(time.time() * 1000)
+            )
+        except AttributeError:
+            # data_point doesn't have time_unix_nano attribute
+            timestamp_ms = int(time.time() * 1000)
+        record.timestamp = timestamp_ms
+
+        # Set attributes
+        try:
+            record.attributes = data_point.attributes
+        except AttributeError:
+            # data_point doesn't have attributes
+            record.attributes = {}
+
+        # For Gauge, set the value directly
+        try:
+            record.value = data_point.value
+        except AttributeError:
+            # data_point doesn't have value
+            record.value = None
+
+        return record
+
+    def _group_by_attributes_and_timestamp(self, record: MetricRecord) -> Tuple[str, int]:
+        """Group metric record by attributes and timestamp.
+
+        Args:
+            record: The metric record
+
+        Returns:
+            A tuple key for grouping
+        """
+        # Create a key for grouping based on attributes
+        attrs_key = self._get_attributes_key(record.attributes)
+        return (attrs_key, record.timestamp)
+
+    def _create_emf_log(
+        self, metric_records: List[MetricRecord], resource: Resource, timestamp: Optional[int] = None
+    ) -> Dict:
+        """
+        Create EMF log dictionary from metric records.
+
+        Since metric_records is already grouped by attributes, this function
+        creates a single EMF log for all records.
+        """
+        # Start with base structure
+        emf_log = {"_aws": {"Timestamp": timestamp or int(time.time() * 1000), "CloudWatchMetrics": []}}
+
+        # Set with latest EMF version schema
+        # opentelemetry-collector-contrib/blob/main/exporter/awsemfexporter/metric_translator.go#L414
+        emf_log["Version"] = "1"
+
+        # Add resource attributes to EMF log but not as dimensions
+        # OTel collector EMF Exporter has a resource_to_telemetry_conversion flag that will convert resource attributes
+        # as regular metric attributes(potential dimensions). However, for this SDK EMF implementation,
+        # we align with the OpenTelemetry concept that all metric attributes are treated as dimensions.
+        # And have resource attributes as just additional metadata in EMF, added otel.resource as prefix to distinguish.
+        if resource and resource.attributes:
+            for key, value in resource.attributes.items():
+                emf_log[f"otel.resource.{key}"] = str(value)
+
+        # Initialize collections for dimensions and metrics
+        metric_definitions = []
+        # Collect attributes from all records (they should be the same for all records in the group)
+        # Only collect once from the first record and apply to all records
+        all_attributes = metric_records[0].attributes if metric_records and metric_records[0].attributes else {}
```

Comment:
```suggestion
        all_attributes = metric_records[0].attributes if metric_records and len(metric_records) > 0 and metric_records[0].attributes else {}
```

Reply:
The current code makes it so that this list is greater than 0. This safety check can be added in next PR.